### PR TITLE
139801559 post to doit with api key v2

### PR DESCRIPTION
**Pull request by Maroo-b**
Addresses: https://www.pivotaltracker.com/story/show/139801559

- add new model for tracing published vol ops to Doit
- add 2 services to interact with Doit API
- add form object to manage posting to Doit
- implement active_job with sucker_punch as backend to avoid extra cost from Heroku

See also https://github.com/AgileVentures/LocalSupport/pull/427

### CODE REVIEW

**Comment by Maroo-b** [comment]
@tansaku after a lot of experiment I managed to figure out the issue with ENV vars.
To be able to execute test with env var, you should submit PR from the same Repo (no PR from Fork is allowed for security reason) check more on [link](https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions)
Also the API Key contains special characters for Bash, which I managed to escape.
Curently the build is failing due to a seo test [link](https://travis-ci.org/AgileVentures/LocalSupport/builds/239764438#L6210)

**Comment by tansaku** [comment]
@Maroo-b that makes sense - great work figuring that out.

**Comment by tansaku** [comment]
@Maroo-b I've closed the other PR from the branch and updated the description for this PR - please check I have it correct

**Comment by tansaku** [comment]
I wish Travis et al had some mechanism to auto-pull out the actual failing tests and add them as comments to the PR like so:

```
  Scenario: Meta Title and Description for volunteers edit page                                  # features/admin/seo.feature:44
    Given I am signed in as a superadmin                                                         # features/step_definitions/authentication_steps.rb:17
    Then I visit the edit page for the volunteer_op titled "Helping Volunteer"                   # features/step_definitions/navigation_steps.rb:89
    Then I should have a page with a title tag set to: "Helping Volunteer | Harrow volunteering" # features/step_definitions/basic_steps.rb:574
      expected "Harrow Community Network | Harrow volunteering" to include "Helping Volunteer | Harrow volunteering" (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/basic_steps.rb:575:in `/^I should have a page with a title tag set to: "([^"]*)"$/'
      features/admin/seo.feature:47:in `Then I should have a page with a title tag set to: "Helping Volunteer | Harrow volunteering"'
    And it should have a Meta Description                                                        # features/step_definitions/basic_steps.rb:578
Failing Scenarios:
cucumber -p second_try features/admin/seo.feature:44 # Scenario: Meta Title and Description for volunteers edit page
```

**Comment by tansaku** [comment]
looks like you have some junk files in the PR `app/controllers/.volunteer_ops_controller.rb.swp` and `app/views/layouts/.two_columns_with_map.html.erb.swp` - I assume this is all up to date with develop.

That error looks strange - something we updated about the titles of the page ...

**Comment by Maroo-b** [comment]
@tansaku sorry for the junk file I'll delete them this evening and I'll rebase this branch with develop in case there is new changes.

**Comment by Maroo-b** [comment]
@tansaku I removed the swap files and made the rebase, what's next ?

**Comment by tansaku** [comment]
@Maroo-b we should replicate these errors locally and get to the bottom of them - I assume you are seeing the same errors locally?

**Comment by Maroo-b** [comment]
@tansaku I added a small fix, I added a dummy test key to make tests pass, and we can share the ENV var for **DOIT_HOST** because it's safe.

**Comment by tansaku** [comment]
thanks @Maroo-b - checking that out now - sorry for delay - totally snowed under
I guess there's no way you can make the weekly meetings on Thursday?  Would there be a good time of the week for you? for hangouts? for text chat?

**Comment by tansaku** [comment]
I have the latest version now running locally ...

File: app/controllers/volunteer_ops_controller.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@ class VolunteerOpsController < ApplicationController
   layout 'two_columns_with_map'
   before_action :set_organisation, only: [:new, :create]
   before_action :authorize, except: [:search, :show, :index]
-  prepend_before_action :set_volunteer_op, only: [:show, :edit]
+  prepend_before_action :set_volunteer_op, only: [:show]
```
Comment (tansaku) [inline app/controllers/volunteer_ops_controller.rb]: @Maroo-b I'd say the change here is causing the error

```
  Scenario: Meta Title and Description for volunteers edit page                                  # features/admin/seo.feature:44
    Given I am signed in as a superadmin                                                         # features/step_definitions/authentication_steps.rb:17
    Then I visit the edit page for the volunteer_op titled "Helping Volunteer"                   # features/step_definitions/navigation_steps.rb:91
    Then I should have a page with a title tag set to: "Helping Volunteer | Harrow volunteering" # features/step_definitions/basic_steps.rb:574
      expected "Harrow Community Network | Harrow volunteering" to include "Helping Volunteer | Harrow volunteering" (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/basic_steps.rb:575:in `/^I should have a page with a title tag set to: "([^"]*)"$/'
      features/admin/seo.feature:47:in `Then I should have a page with a title tag set to: "Helping Volunteer | Harrow volunteering"'
    And it should have a Meta Description                                                        # features/step_definitions/basic_steps.rb:578

```

Comment (tansaku) [inline app/controllers/volunteer_ops_controller.rb]: the volunteer_op needs to be set on the edit page to support the SEO tag

**Comment by Maroo-b** [comment]
@tansaku thank you for the feedback and sorry fo the late reply.
Unfortunately, I can't attend weekly meetings on Thursday. By the way is this PR good to merge so I can start editing the rake task for importing doit vol ops ( to avoid importing posted vol ops from LocalSupport).

Thank you.

**Comment by tansaku** [comment]
thanks @Maroo-b - are their other days/times you could attend?

BTW, before we merge this do you think we want to fix up any of the codeclimate issues?

**Comment by Maroo-b** [comment]
@tansaku absolutely, I sent a commit to fix some codeclimate issues. I'll send you a DM for details.

**Comment by tansaku** [comment]
@Maroo-b any updates on this PR?

**Comment by tansaku** [comment]
so I think I've got some more codeclimate fixes here: 

https://github.com/AgileVentures/LocalSupport/pull/468

and I get an option to post to doit here - which is great:

screenshot 2017-08-12 11 07 05

but then I guess I need to set up some keys locally to have any organisations listed.

I'd love to try out with our AgileVentures account.  Have you any idea how I get some set of managed organisations.  At the moment we are using our API key and I guess that's not yet associated with any organisations?

p.s. I also notice the form layout is a bit weird

**Comment by tansaku** [comment]
urgh, although looks like my attempts to fix remaining code climate issues has not been so successful ...  :-(

**Comment by Maroo-b** [comment]
@tansaku thank you for fixing the remaining codeclimate issues. (I thought that the PR was ready to merge)
Regarding the list of organisations, I suggest that we use the staging API Key for the moment. I tried before to create an account on staging but it seems that the account validation via email is disabled. 

The issue of the form is related to the migration to bootstrap 3, I guess it will be better to fix it separately what do you think?

**Comment by tansaku** [comment]
@Maroo-b agreed the form layout is a different issue.

I think the issue with the staging API key is that it's not associated with any orgs on do-it - or have you seen otherwise? 

Without associated orgs we can't really test fully can we?  I've emailed do-it to ask them to add our AV org to the API keys - we'll see what they say

**Comment by tansaku** [comment]
@Maroo-b I got the keys working:

screenshot 2017-08-15 10 13 32

the issue for me locally was that I had `X-AUTH-TOKEN` in application.yml rather than `DOIT_AUTH_TOKEN`

**Comment by tansaku** [comment]
@Maroo-b I just succesfully posted a volunteer op to doit from my local dev instance:

https://do-it.org/opportunities/80b2477e-d49d-409d-a09d-ea65bfb1e895

outstanding work on your part - sorry it took me to long to get my review in - really wonderful piece of work from you here - ready to merge - I just notice some conflict in the schema.rb - gonna fix that myself and get this merged in ...

**Comment by tansaku** [comment]
ah, and I'm just thinking if we publish to do-it, then we have the issue that we don't necessarily want to store it locally - or we want to avoid pulling in the op from do-it to avoid doubling up with a local store of the op ...

**Comment by tansaku** [comment]
related follow up work:

https://www.pivotaltracker.com/story/show/150247799
https://www.pivotaltracker.com/story/show/150247828

**Comment by Maroo-b** [comment]
@tansaku thank you for checking the PR, 
I'll start with the duplication ticket, I already anticipated it with DoitTrace model: https://github.com/AgileVentures/LocalSupport/pull/453/files#diff-d835f610f4bf26607772d333580de6c1

**Pull request merged**